### DISCUSSION TITLE
Returning the results of juniper::execute() in a context function

### DISCUSSION THREAD

**Discussion by twiclo** [graphql-rust/juniper#1109] [Q&A] [answered]
I want my users to be able to say "Give me the results of query x" where they would provide me an object to specify which query they want ran. I can take that ID, get the hashed query from my DB, then run that using `junpier::execute()`. Only problem is I don't know how I would actually return that to the user.

Here's my `Query` implement:

```
 pub struct Context {
   pool: sqlx::PgPool,
 }
 impl juniper::Context for Context {}
 struct Query;

 type Schema = RootNode<'static, Query, Mutation, EmptySubscription<Context>>;

 fn schema() -> Schema {
   Schema::new(Query, EmptyMutation, EmptySubscription::<Context>::new())
 }

 #[graphql_object(context = Context)]
 impl Query {
   fn apiVersion() -> &'static str {
     "1.0"
   }

   async fn user(ctx: &Context, id: i32) -> FieldResult<User> {
     Ok(User::get_own_user(validate_perm(&ctx.user.perms)?, ctx).await?)
   }

   async fn query(ctx: &Context, id: i32) -> Result<NoIdea> {
     Ok(get_query_from_db(ctx, id))
   }
 }
```

Can anyone give me some guidance on what I should return here? I doubt it'd be a `FieldResult` since I'm trying to return the entire query

**Comment by tyranron** [comment, marked as answer]
@twiclo you want here some sort of dynamic values (not known statically), but GraphQL schema is typed. You, either way, should specify all the possible concrete types returned by `query` resolver. They could be united either under a GraphQL Interface, or under a GraphQL Union. Another way is to use some custom opaque GraphQL scalar type, but in this case you won't be able to interact with its inners inside GraphQL.

Could you describe with a raw GraphQL Schema Language how your schema should look like?

**Reply by twiclo** [reply]
"Could you describe with a raw GraphQL Schema Language how your schema should look like?"
No not really. Saved query 1 could be something like
```
{
  teams {
    name,
    members {
      username
      name {
        first,
        lastf
      }
    }
  }
}
```
While query 2 could be
```
{
  tickets {
    title,
    description
  }
}
```

I'm trying to mimic the idea of a "persisted query" described here: https://www.apollographql.com/blog/apollo-client/persisted-graphql-queries/

**Reply by tyranron** [reply]
@twiclo your persisted queries still should be typed. So, as far as you need to interact with them in GraphQL, you have no choice but to describe in the GraphQL schema all the possible results, then unite them under a single GraphQL union, and use that union as the returned type of the `query` resolver.

**Reply by twiclo** [reply]
So when you say I need to unite these under a GraphQL union are you telling me I need to have an enum of possible return types that derives `GraphQLUnion`? and all I would need to do is return that enum?

**Reply by tyranron** [reply]
@twiclo yes

**Reply by twiclo** [reply]
Thanks for your help. I decided to just create a new REST endpoint that'll take a query ID and respond with the returned graphql query. Only problem with that is I get back a bunch of `juniper::Value` types. Is there an easy way to convert this to json or will I need to do that manually?

```
[src/handlers/user_handlers/queries.rs:49] juniper::execute(&query, None, &schema, &Variables::new(),
            &ctx).await.unwrap() = (
    Object(
        Object {
            key_value_list: {
                "users": List(
                    [
                        Object(
                            Object {
                                key_value_list: {
                                    "username": Scalar(
                                        String(
                                            "twilcox",
                                        ),
                                    ),
                                },
                            },
                        ),
                        Object(
                            Object {
                                key_value_list: {
                                    "username": Scalar(
                                        String(
                                            "bschlob",
                                        ),
                                    ),
                                },
                            },
                        ),
                    ],
                ),
            },
        },
    ),
    [],
)
```

**Reply by tyranron** [reply]
@twiclo `juniper::Value` implements `serde::Serialize`, so `serde_json::to_string(&value)` should do the job.
